### ISSUE TITLE
Language changes not available to assistive technologies

### ISSUE THREAD

**Issue by AlainVagner** [mozilla/pdf.js#14261]
Attach (recommended) or Link to PDF file here:
[test-lang.pdf](https://github.com/mozilla/pdf.js/files/7520359/test-lang.pdf)
The default language in this document is english. But the document contains 2 language changes.

Configuration:
- Web browser and its version: Firefox 93.0
- Operating system and its version: Windows 10 Version 1909 Build 18363.1256
- PDF.js version: from the master branch
- Is a browser extension: no
- Screen reader: NVDA 2021.2
- OS language: english

Steps to reproduce the problem:
1. Open the file in PDF.js
2. Read the file with NVDA
3. The text in French is read with an English voice (in Acrobat, the text in French is read with a French voice)

What is the expected behavior? (add screenshot)
French text should be read with a French voice.

In this PDF document, the content in French is made of two tags, one H1 and one P. These two tags have the property Language set to "French". This info should be available as a `lang` attribute in the DOM set to "French", so that screen readers like NVDA can select the correct language to read this text.
![screen shot of a language change in Acrobat Pro](https://user-images.githubusercontent.com/16536731/141300640-e8c91329-a77c-400d-89b4-68d066c33db9.png)

What went wrong? (add screenshot)
Here the only lang attribute we have in the DOM is the one of the default language. We should find 2 others, one for the title in French and one for the paragraph in French.
<img width="1438" alt="Screenshot of the dev tools showing only one alt attribute" src="https://user-images.githubusercontent.com/16536731/141300905-01e74920-7066-4136-8814-92b26735d337.png">

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):
/

**Comment by Snuffleupagus** [comment]
Can you please check if the preview in https://github.com/mozilla/pdf.js/pull/14262#issuecomment-966313769 is enough to address this?

**Comment by AlainVagner** [comment]
Thanks for this swift answer! I tested it and it works in Chrome, but for some reason it doesn't in Firefox (I tested with 2 screen readers, NVDA and Jaws). I wonder if it is related to the use of `aria-owns`.

**Comment by Snuffleupagus** [comment]
@brendandahl Any ideas about the Firefox-specific issue mentioned in https://github.com/mozilla/pdf.js/issues/14261#issuecomment-966355837?

**Comment by AlainVagner** [comment]
Just in case it could help, I have made a small test case for this issue: 
https://codepen.io/alainvagner/full/GRvBaBB

**Comment by AlainVagner** [comment]
By the way, `aria-owns` is not supported by VoiceOver on Mac and iOS, according to https://a11ysupport.io/tech/aria/aria-owns_attribute I have made some tests in Firefox 93 on MacOS 11.6, and I was not able to read this document at all via pdf.js. Does pdf.js support Firefox on MacOS + VoiceOver?

**Comment by brendandahl** [comment]
@jcsteh Do you know why we have different behavior in chrome and firefox for the test case in [this comment](https://github.com/mozilla/pdf.js/issues/14261#issuecomment-966409238).

**Comment by jcsteh** [comment]
> By the way, `aria-owns` is not supported by VoiceOver on Mac and iOS, according to https://a11ysupport.io/tech/aria/aria-owns_attribute

Firefox on Mac exposes aria-owns by altering the actual structure of the a11y tree such that it should be transparent to VoiceOver that aria-owns is being used. That is, VoiceOver should see the same tree regardless of whether it was structured using the DOM or aria-owns. There could be bugs here, of course, but I'm not aware of any currently.

> I have made some tests in Firefox 93 on MacOS 11.6, and I was not able to read this document at all via pdf.js. Does pdf.js support Firefox on MacOS + VoiceOver?

That's concerning. Was this with Firefox's internal pdf.js? Would you mind filing a separate bug for that?

**Comment by jcsteh** [comment]
> @jcsteh Do you know why we have different behavior in chrome and firefox for the test case in [this comment](https://github.com/mozilla/pdf.js/issues/14261#issuecomment-966409238).

Interesting. Language is exposed as a text attribute. All text attributes are calculated using DOM/layout info only. Normally, that makes sense; I don't think we'd want to inherit font information from an ARIA owner, for example, as that would override the actual font info of the owned text. There's perhaps an argument for looking at the a11y tree for language because language is more semantic than presentational. I think this is a grey area; I don't know of any spec that covers this case.

I wonder if Chromium did this deliberately to support this case or whether it was more of a happy accident. I'm guessing the latter. Chromium also does this for superscript, for example, which seems potentially problematic to me.

I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1741122 against the Gecko a11y engine and we'll see where we land.

**Comment by AlainVagner** [comment]
> Firefox on Mac exposes aria-owns by altering the actual structure of the a11y tree such that it should be transparent to VoiceOver that aria-owns is being used. That is, VoiceOver should see the same tree regardless of whether it was structured using the DOM or aria-owns. There could be bugs here, of course, but I'm not aware of any currently.
> 
I didn't know that, thanks for the info!

> That's concerning. Was this with Firefox's internal pdf.js? Would you mind filing a separate bug for that?
I tested this with Firefox 93 and pdf.js with the fixes discussed here. I will give it another try with a more recent version of Firefox.

**Comment by AlainVagner** [comment]
> I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1741122 against the Gecko a11y engine and we'll see where we land.
@jcsteh thanks a lot!

**Comment by jcsteh** [comment]
@brendandahl, is #14262 a complete fix for this issue? It's unclear that cascading the lang attribute via aria-owns is the right thing to do, and if there's already a complete fix landed here, it probably doesn't make sense for us to change this at this point. Thanks.

**Comment by Snuffleupagus** [comment]
> is #14262 a complete fix for this issue?

On its own that one won't do anything, unfortunately.
That PR was landed under the *assumption* that [bug 1741122](https://bugzilla.mozilla.org/show_bug.cgi?id=1741122) will add the necessary functionality in Firefox; please also see https://github.com/mozilla/pdf.js/pull/14262#issuecomment-968813144

**Comment by jcsteh** [comment]
My apologies. I misinterpreted the PR as being a fix for pdf.js language exposure in Firefox, rather than language exposure in general. I should have read the earlier comments here more closely. I'll continue to look into how we can manage this in the Gecko a11y engine.

**Comment by AlainVagner** [comment]
From my end user perspective, the goal is to be able to have language exposure for pdf files in Firefox. 
I am wondering if it would not be easier to fix this issue in the markup in PDF.js than changing the accessibility engine in Gecko. Would this be possible? If needed I can make some proposals.

**Comment by Snuffleupagus** [comment]
> Would this be possible? If needed I can make some proposals.

It's not really a question about possibility, as far as I'm concerned, but more the fact that it'd *not* be an entirely desirable change from the general PDF.js library perspective (unless it turns out to be the only available solution).

Please note that currently the text-extraction and structureTree-building are *completely separate* operations (in the worker-thread), which can be run in any order (or even in parallel). If we were to instead encode the `/Lang`-attribute into the textLayer itself, that'd mean that the structureTree would always need to be created *before* the text-extraction can start.

**Comment by jcsteh** [comment]
That clarifies things. Relative to that, this should be easy enough to fix
in Gecko. I just need to find the time to write the patch.

**Comment by anton-mauritzson** [comment]
@jcsteh did this ever get resolved, i'm struggling to find a solution for this exact problem.
